### PR TITLE
Fixed voices and filter functions with invalid range checking

### PR DESCRIPTION
**Pull request by gitlarryf**

### CODE REVIEW

File: easysid.cpp [hunk 1]
```diff
@@ -65,10 +65,10 @@ EXPORT void set_pw(uint8_t voice, uint16_t pw)
 
 EXPORT void set_wave(uint8_t voice, uint8_t wave)
 {
-    if (voice < 1 || voice > 3) {
+    if (voice < 0 || voice > 2) {
         return;
     }
-    if (wave >= 4) {
+    if (wave > 3) {
```
Comment (ghewgill) [inline easysid.cpp]: What's the purpose of this change? Since wave is an integer, these statements are equivalent, aren't they?

File: easysid.cpp [hunk 2]
```diff
@@ -152,42 +152,58 @@ EXPORT void set_filter_res(uint8_t res)
     sid.write(23, reg);
 }
 
-EXPORT void set_filter_enable(uint8_t voice)
+EXPORT void set_filter(uint8_t voices, bool enable)
```
Comment (ghewgill) [inline easysid.cpp]: Don't use a bit mask for this. There's no reason to complicate the interface that way. If the caller needs to enable the filter on more than one voice, then call this function more than once.

File: easysid.cpp [hunk 3]
```diff
@@ -204,7 +220,7 @@ EXPORT void run_ms(uint32_t ms)
     if (pcm == NULL) {
         pcm = makePcmOutputAudio();
     }
-    const size_t buflength = 16384;
+    const size_t buflength = 32768;
```
Comment (ghewgill) [inline easysid.cpp]: What's this change for? It doesn't seem to be related to validation of input parameters. What's the justification here?

File: easysid.cpp (hunk 1 above)
Comment (gitlarryf) [inline easysid.cpp]: This was done to match the rest of the file, and the range checking it is doing.  This just makes everything more consistent.

File: easysid.cpp [hunk 4]
```diff
@@ -152,42 +152,66 @@ EXPORT void set_filter_res(uint8_t res)
...
+EXPORT void set_filter(uint8_t voice, bool enable)
 {
-    if (voice < 1 || voice > 3) {
+    if (voice < 0 || voice > 2) {
         return;
     }
     char reg = sid.read_state().sid_register[23];
-    reg &= ~0x0f;
-    reg |= 1 < (voice - 1);
+    reg &= ~(1 << voice) & 0x0f;
```
Comment (ghewgill) [inline easysid.cpp]: I'm pretty sure this doesn't do what you want it to do. Check the precedence of `~` and `&`. Make sure you don't change bits that you don't want to change.

File: easysid.cpp [hunk 5]
```diff
@@ -152,42 +157,66 @@ EXPORT void set_filter_res(uint8_t res)
...
-EXPORT void set_filter_enable(uint8_t voice)
+EXPORT void set_filter(uint8_t voice, bool enable)
 {
     if (voice < 1 || voice > 3) {
         return;
     }
     char reg = sid.read_state().sid_register[23];
-    reg &= ~0x0f;
-    reg |= 1 < (voice - 1);
+    reg &= ~1 << (voice - 1);
```
Comment (ghewgill) [inline easysid.cpp]: This still isn't right. The above expression does not give the intended results. The precedence rules of C++ are here for reference: http://en.cppreference.com/w/cpp/language/operator_precedence

File: easysid.cpp [hunk 6]
```diff
@@ -152,42 +157,66 @@ EXPORT void set_filter_res(uint8_t res)
...
+    if (enable) {
+        reg |= (1 << voice - 1);
+    }
+    sid.write(23, reg);
+}
+
+EXPORT void set_filter_external(bool enable)
+{
+    char reg = sid.read_state().sid_register[23];
+    reg &= ~(1 << 3) & 0x0f;
```
Comment (ghewgill) [inline easysid.cpp]: And as discussed, this is still wrong. You're only modifying a single bit here, it should look pretty much the same as the other functions that modify a single bit.

File: easysid.cpp [hunk 7]
```diff
@@ -152,42 +157,66 @@ EXPORT void set_filter_res(uint8_t res)
...
-EXPORT void set_filter_enable(uint8_t voice)
+EXPORT void set_filter(uint8_t voice, bool enable)
 {
     if (voice < 1 || voice > 3) {
         return;
     }
     char reg = sid.read_state().sid_register[23];
-    reg &= ~0x0f;
-    reg |= 1 < (voice - 1);
+    reg &= ~(1 << (voice - 1));
```
Comment (ghewgill) [inline easysid.cpp]: If you're going to use the value `(1 << (voice - 1))` twice, make that calculation just one and save it in a local variable called `mask` or something.

File: easysid.cpp [hunk 8]
```diff
@@ -152,42 +157,68 @@ EXPORT void set_filter_res(uint8_t res)
...
+    if (enable) {
+        reg |= mask;
+    }
+    sid.write(23, reg);
+}
+
+EXPORT void set_filter_external(bool enable)
+{
+    char reg = sid.read_state().sid_register[23];
+    reg8 mask = (1 << 3);
```
Comment (ghewgill) [inline easysid.cpp]: You don't need to make a local variable for this. It's a constant. It equals `0x08`. Just use that value instead of `mask`, similar to how the following functions work.
